_posts/2019-08-26-how-to-use-gruntwork-service-catalog.adoc

```diff
@@ -0,0 +1,1934 @@
+deploy AMIs across one or more servers (e.g., into an AWS Auto Scaling Group), with support for auto scaling, auto
+healing, zero-downtime deployments, etc.
+
+The Gruntwork Service Catalog contains a number of scripts and binaries that you can run on your servers: e.g., the
+ELK code includes scripts you run during boot on Elasticsearch servers to bootstrap the cluster, and the security code
+includes an `ssh-grunt` binary you can run on each server to manage SSH access to that server using IAM groups (i.e.,
+IAM users in specific IAM groups will be able to SSH to specific servers using their own usernames and SSH keys).
+
+To get these scripts and binaries onto your virtual servers (e.g., onto EC2 instances in AWS or compute instances in
+GCP), we recommend using Packer to build VM images that have these scripts and binaries installed. You'll see an
+example of how to do this in <<how_to_use_the_catalog>>. Note that Gruntwork Service Catalog does NOT require that
+you use Packer (e.g., you could also use Ansible or Chef to install the scripts and binaries), but the Gruntwork
+Reference Architecture does use Packer as one of its opinionated tools.
+
+[[docker]]
+=== Docker
+
+https://www.docker.com[Docker] is an open source tool you can use to run _containers_ and define _container images_ as
+code. A container is a bit like a lightweight VM, except instead of virtualizing all the hardware and the entire
+operating systems, containers virtualize solely user space, which gives you many of the isolation benefits of a VM
+(each container is isolated in terms of memory, CPU, networking, hard drive, etc), but with much less memory, CPU, and
+start-up time overhead. For example, here is how you can define an Ubuntu 18.04 Docker image that has Node.js installed:
+
+.Dockerfile
+[source,Dockerfile]
+----
+FROM ubuntu:18.04
+
+RUN curl -sL https://deb.nodesource.com/setup_10.x | sudo -E bash - && \
+    sudo apt-get update -y && \
+    sudo apt-get install -y nodejs
+----
+
+You can run `docker build -t example-image .` to build a Docker image from this code, push the image to a Docker
+Registry (e.g., ECR or Docker Hub), and then deploy the Docker image using other tools. For example, the Gruntwork
+Service Catalog contains a number of modules for running _container orchestration tools_ such as Kubernetes, ECS, and
+Nomad that you can use to deploy and manage Docker images.
+
+[[immutable_infrastructure]]
+=== Immutable infrastructure
+
+With _mutable infrastructure_, you deploy a set of servers, and you continuously update those servers in place. Every
+new update gets installed on top of the previous updates, either manually (e.g., by SSHing to each server and running
+commands), or via tools like Ansible, Chef, or Puppet. Over time, each "mutable" server builds up a history of
+changes, which can make it difficult to (a) reason about what's actually installed and (b) debug issues that are
+specific to the unique history of one server but not others.
```

Comment:
Also servers that were meant to be equal will eventually diverge.

Reply:
Updated: https://github.com/gruntwork-io/gruntwork-io.github.io/pull/151/commits/4603b48a336569db90352172790bc29347291c75